gcloud-java-core/src/main/java/com/google/gcloud/BaseWriteChannel.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud;
+
+import com.google.common.base.MoreObjects;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Base implementation for a {@link WriteChannel}.
+ *
+ * @param <ServiceOptionsT> the service options used by the channel to issue RPC requests
+ * @param <EntityT> the entity this channel writes data to. Possibly with additional configuration
+ */
+public abstract class BaseWriteChannel<
+    ServiceOptionsT extends ServiceOptions<?, ?, ServiceOptionsT>,
+    EntityT extends Serializable> implements WriteChannel {
+
+  private static final int MIN_CHUNK_SIZE = 256 * 1024;
+  private static final int DEFAULT_CHUNK_SIZE = 8 * MIN_CHUNK_SIZE;
+
+  protected final ServiceOptionsT options;
+  protected final EntityT entity;
+  protected final String uploadId;
+  protected int position;
+  protected byte[] buffer = new byte[0];
+  protected int limit;
+  protected boolean isOpen = true;
+  protected int chunkSize = defaultChunkSize();
+
+  protected int minChunkSize() {
+    return MIN_CHUNK_SIZE;
+  }
+
+  protected int defaultChunkSize() {
+    return DEFAULT_CHUNK_SIZE;
+  }
+
+  protected abstract void write(int length, boolean last);
+
+  protected BaseWriteChannel(ServiceOptionsT options, EntityT entity, String uploadId) {
+    this.options = options;
+    this.entity = entity;
+    this.uploadId = uploadId;
+  }
+
+  protected void flush() {
```

Comment:
I would make anything that is not need to be overriden by the subclasses either package scope, private or final (if must be public). Especially methods that would easily get it wrong and require deep understanding on the base class internals.